### DISCUSSION TITLE
LSP setup support for the Flow language

### DISCUSSION THREAD

**Discussion by maxclaus** [mason-org/mason.nvim#1403] [Q&A]
I am trying to get [Flow](https://flow.org/) LSP support to work with NeoVim, so it can have proper type checking for flow files.
It seems though, mason does not have out of the box configuration as it does for typescript, and I am having a hard time finding examples anywhere else on how to get it working. So, I was hoping I could get some guidance on how to achieve that.

I noticed that nvim-lspconfig does have flow support: https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/flow.lua.

But when I try to use Mason to ensure the Flow LSP is installed, as it does for typescript, I get this error:

```
[mason-lspconfig.nvim] Server "flow" is not a valid entry in ensure_installed. Make sure to only provide lspconfig server names.
```

I tried to add flow to https://github.com/williamboman/mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/mappings/server.lua, but I still get that same error. 

And I get this error `"flow" is not a valid package` if I try to run `:MasonInstall flow` or `:LspInstall flow` too.

Appreciate any guidance, thanks.

**Comment by maxclaus** [comment]
Another thing I found, it needs to use tree-sitter-typescript for Flow files with `javascriptreact` filetype, that way the tree-sitter-typescript can handle Flow files https://github.com/nvim-treesitter/nvim-treesitter/issues/3350

**Comment by williamboman** [comment]
Hello! This is because flow doesn't exist as a package in Mason. Generally this is a package that gets installed locally in each project it's used.

**Comment by maxclaus** [comment]
Hi William. I am confused though because `mason-lspconfig` does have `tsserver` support, the `ensure_installed` does work for `tssever`. So, I just imagined it should work similarly to the Flow LSP server.

I got the type checking and syntax highlight working for Flow though without using mason-lspconfig. So, I was just curious why it works differently for tsserver as it seemed flow lsp server and the tsserver lsp should have a similar setup and support.

**Reply by williamboman** [reply]
It will only work for packages available via Mason. `tsserver` corresponds to the `typescript-language-server` Mason package, and `flow` doesn't have a corresponding package in Mason, so that's why. You can browse the full list of available packages [here](https://mason-registry.dev/registry/list)

**Reply by maxclaus** [reply]
I see, I am sorry I think I misread your previous comment, so we would need to get the Flow LSP added as a package first. 

I was checking `nvim-lspconfig` back again and noticed that the [ts-server configuration](https://github.com/neovim/nvim-lspconfig/blob/ba74226cac83eacb435827e028f810212147937b/lua/lspconfig/server_configurations/tsserver.lua#L6) is using the `typescript-language-server` from the path like this `typescript-language-server --stdio`, which makes sense have it as a package in Mason. But for the [flow configuration](https://github.com/neovim/nvim-lspconfig/blob/ba74226cac83eacb435827e028f810212147937b/lua/lspconfig/server_configurations/flow.lua#L5) it uses `npx --no-install flow lsp` instead, which makes sense because it is using the Flow version the current project is using.

So, it seems Flow shouldn't be added to Mason. Do you think that makes sense?

**Reply by williamboman** [reply]
> So, it seems Flow shouldn't be added to Mason. Do you think that makes sense?

Yeah! This is why it hasn't been added yet
